### PR TITLE
Align accuracy for MAE finetune

### PR DESCRIPTION
**Pull request by leaves-zwx**

### CODE REVIEW

File: projects/MAE/configs/mae_finetune.py [hunk 1]
```diff
@@ -5,17 +5,18 @@
...
 # Path to the weight for fine-tune
 finetune = OmegaConf.create()
 finetune.enable = True  # only load weight if enable is True
 finetune.weight_style = (
-    "oneflow"  # Set "oneflow" for loading oneflow weights, set "pytorch" for loading torch weights
+    # "oneflow"  # Set "oneflow" for loading oneflow weights, set "pytorch" for loading torch weights
+    "pytorch"
 )
-finetune.path = "/path/to/pretrained_mae_weight"
+finetune.path = "/home/zhangwenxiao/wksp/mae/mae_pretrain_vit_base.pth"
```
Comment (leaves-zwx) [inline projects/MAE/configs/mae_finetune.py:19]: 目前手头上还没训好的 pretrain 模型，所以使用 https://github.com/facebookresearch/mae 中提供好的 pretrain 模型。

File: projects/MAE/configs/mae_finetune.py [hunk 2]
```diff
@@ -26,12 +27,11 @@
...
 # Refine data path to imagenet
-dataloader.train.dataset[0].root = "/path/to/imagenet"
-dataloader.test[0].dataset.root = "/path/to/imagenet"
-
+dataloader.train.dataset[0].root = "/ssd/dataset/ImageNet/extract"
+dataloader.test[0].dataset.root = "/ssd/dataset/ImageNet/extract"
 
 # Graph training
-graph.enabled = False
+graph.enabled = True
```
Comment (leaves-zwx) [inline projects/MAE/configs/mae_finetune.py:34]: 打开 graph 来调试，因为目前 eager 下会慢一些，希望能更快看到更多 epoch 的 accuracy 来验证调整的效果。

File: projects/MAE/configs/mae_finetune.py [hunk 3]
```diff
@@ -50,49 +50,67 @@
...
+n_gpus = 8
+
+# dataset length
+dataset_train_length = 1281167
+dataset_val_length = 50000
+
 # Refine training settings for MAE finetune
-train.train_micro_batch_size = 128
+train.train_micro_batch_size = 32
+train.num_accumulation_steps = 4
```
Comment (leaves-zwx) [inline projects/MAE/configs/mae_finetune.py:62]: 我调试使用的是 3080ti 12G 显存的卡，跑不起来的 128 的 batch size，所以用 accumulation steps 来匹配。

File: projects/MAE/configs/mae_finetune.py [hunk 4]
```diff
@@ -50,49 +50,67 @@
...
 train.warmup_ratio = 5 / 100
-train.log_period = 1
-train.evaluation.eval_period = 1000
+train.log_period = 20
+train.evaluation.eval_period = epoch_iter
+train.checkpointer.period = epoch_iter
 
 # Set layer decay for MAE fine-tune
-train.layer_decay = 0.75
+train.layer_decay = 0.65
```
Comment (leaves-zwx) [inline projects/MAE/configs/mae_finetune.py:75]: 对齐 https://github.com/facebookresearch/mae/blob/efb2a8062c206524e35e47d04501ed4f544c0ae8/FINETUNE.md?plain=1#L121

File: projects/MAE/configs/mae_finetune.py [hunk 5]
```diff
@@ -50,49 +50,67 @@
...
+
+# AMP
+train.amp.enabled = True
```
Comment (leaves-zwx) [inline projects/MAE/configs/mae_finetune.py:78]: 开启 amp

File: projects/MAE/configs/mae_finetune.py [hunk 6]
```diff
@@ -50,49 +50,67 @@
...
 
 # Base learning in MAE is set to 1.5e-4
 # The actually learning rate should be computed by linear scaling rule as follows:
 # lr = base_lr * batch_size / 256
 # In LiBai, you should refine the actually learning rate due to your on settings
 # Here we use 8 GPUs, 128 batch_size per GPU for training, batch_size equals to 1024
-base_lr = 1e-3
-actual_lr = base_lr * (train.train_micro_batch_size * 8 / 256)
-
+base_lr = 5e-4
```
Comment (leaves-zwx) [inline projects/MAE/configs/mae_finetune.py:85]: 对齐 https://github.com/facebookresearch/mae/blob/efb2a8062c206524e35e47d04501ed4f544c0ae8/FINETUNE.md?plain=1#L121

File: projects/MAE/configs/mae_finetune.py [hunk 7]
```diff
@@ -50,49 +50,67 @@
...
-train.scheduler._target_ = warmup_layerscale_cosine_lr_scheduler
-train.scheduler.warmup_factor = 0.001
-train.scheduler.alpha = 0.0
-train.scheduler.warmup_method = "linear"
-
+if graph.enabled:
+    train.scheduler = LazyCall(warmup_cosine_lr_scheduler)(
+        warmup_factor=0.0,
+        min_lr=1e-6,
+    )
```
Comment (leaves-zwx) [inline projects/MAE/configs/mae_finetune.py:105]: graph 模式下，layer scale 是内置实现，只需要在 optimizer 的 param_group 里面设置好 `lr_scale` 即可。

File: projects/MAE/configs/mae_finetune.py [hunk 8]
```diff
@@ -50,49 +50,67 @@
...
+else:
+    train.scheduler = LazyCall(warmup_layerscale_cosine_lr_scheduler)(
+        warmup_factor=0.0,
+        min_lr=1e-6,
+    )
```
Comment (leaves-zwx) [inline projects/MAE/configs/mae_finetune.py:110]: 该处实现略微调整

File: projects/MAE/train_net.py [hunk 9]
```diff
@@ -24,16 +24,16 @@
...
     @classmethod
     def build_model(cls, cfg):
-        if try_get_key(cfg, "graph.enabled") is True:
-            raise NotImplementedError(
-                "LiBai MAE only support eager global mode now, please set cfg.graph.enabled=False"  # noqa
-            )
+        # if try_get_key(cfg, "graph.enabled") is True:
+        #     raise NotImplementedError(
+        #         "LiBai MAE only support eager global mode now, please set cfg.graph.enabled=False"  # noqa
+        #     )
```
Comment (leaves-zwx) [inline projects/MAE/train_net.py:36]: 只要支持 layer scale 功能后，graph 下也可运行

File: projects/MAE/utils/lr_decay.py [hunk 10]
```diff
@@ -45,6 +48,8 @@ def param_groups_lrd(model, weight_decay=0.05, layer_decay=0.75):
         layer_idx = get_layer_idx_for_vit(name, num_layers)
         group_name = "layer_%d_%s" % (layer_idx, g_decay)
 
+        # logger.info(f"{name}, shape={param.shape}, {g_decay}={this_decay}, layer_scale={layer_scales[layer_idx]}")
```
Comment (leaves-zwx) [inline projects/MAE/utils/lr_decay.py:51]: 调试，查看 weight decay 和 layer scale 是否如预期。

File: projects/MAE/utils/scheduler.py [hunk 11]
```diff
@@ -17,63 +17,90 @@
...
 
 import oneflow as flow
-from oneflow.optim.lr_scheduler import CosineDecayLR
+from oneflow.optim.lr_scheduler import _LRScheduler
 
 logger = logging.getLogger(__name__)
 
 
-class WarmupLayerScaleCosineDecayLR(CosineDecayLR):
+class LayerScaleWarmupCosineDecayLR(_LRScheduler):
```
Comment (leaves-zwx) [inline projects/MAE/utils/scheduler.py:25]: 在 WarmupCosineDecayLR 的基础上增加了 layer scale 的功能，但由于 oneflow graph 模式暂未支持自定义 lr scheduler 功能，所以目前只能在 eager 下使用。

File: projects/MAE/utils/scheduler.py [hunk 12]
```diff
@@ -17,63 +17,90 @@
...
+    warmup_iter: int,
     warmup_factor: float,
+    min_lr: float = 0.0,
+):
+    return LayerScaleWarmupCosineDecayLR(
+        optimizer, steps=max_iter, warmup_steps=warmup_iter, warmup_factor=warmup_factor, min_lr=min_lr
+    )
+
+
+def warmup_cosine_lr_scheduler(
```
Comment (leaves-zwx) [inline projects/MAE/utils/scheduler.py:82]: 改成内置的 CosineAnnealingLR + WarmupLR 来实现，这样 graph 下也可支持该 scheduler，同时 layer scale 功能在 graph 下不用 lr scheduler 来负责进行 scale，而是由 optimizer 设置好 lr_scale 即可。

File: projects/MAE/utils/weight_convert.py [hunk 13]
```diff
@@ -80,10 +82,16 @@ def filter_keys(key, value, cfg):
         key = key.replace("mlp.fc2", "mlp.dense_4h_to_h")
     elif "fc_norm" in key:
         key = key.replace("fc_norm", "norm")
+    elif key == "norm.weight" or key == "norm.bias":
+        value = None
```
Comment (leaves-zwx) [inline projects/MAE/utils/weight_convert.py:86]: 对齐 https://github.com/facebookresearch/mae/blob/be47fef7a727943547afb0c670cf1b26034c3c89/models_vit.py#L27-L32

global_pool 时，会丢弃掉 pretrain 模型中最后的 norm 的参数。

File: projects/MAE/utils/weight_convert.py [hunk 14]
```diff
@@ -80,10 +82,16 @@ def filter_keys(key, value, cfg):
...
 
     return key, value
 
 
+def log_param(key, value):
+    logger.info(f"{key}, shape={value.shape}")
```
Comment (leaves-zwx) [inline projects/MAE/utils/weight_convert.py:92]: 调试功能

File: projects/MAE/utils/weight_convert.py [hunk 15]
```diff
@@ -94,14 +102,19 @@ def load_torch_checkpoint(model, cfg, path="./mae_finetuned_vit_base.pth", stric
     parameters = torch_dict
     new_parameters = dict()
     for key, value in parameters.items():
+        # log_param(key, value)
         if "num_batches_tracked" not in key:
             # to global tensor
             key, val = filter_keys(key, value, cfg)
+            if val is None:
+                continue
```
Comment (leaves-zwx) [inline projects/MAE/utils/weight_convert.py:110]: 遇到被丢弃的参数，则跳过

File: projects/MAE/utils/weight_convert.py [hunk 16]
```diff
@@ -94,14 +102,19 @@ def load_torch_checkpoint(model, cfg, path="./mae_finetuned_vit_base.pth", stric
...
             val = val.detach().cpu().numpy()
             val = flow.tensor(val).to_global(
                 sbp=flow.sbp.broadcast, placement=flow.placement("cuda", ranks=[0])
             )
             new_parameters[key] = val
-    model.load_state_dict(new_parameters, strict=strict)
-    print("Successfully load torch mae checkpoint.")
+    msg = model.load_state_dict(new_parameters, strict=strict)
+    logger.info(msg)
+    trunc_normal_(model.head.weight, std=2e-5)
```
Comment (leaves-zwx) [inline projects/MAE/utils/weight_convert.py:118]: 对齐 https://github.com/facebookresearch/mae/blob/be47fef7a727943547afb0c670cf1b26034c3c89/main_finetune.py#L257

File: projects/MAE/configs/mae_finetune.py (hunk 1 above)
Comment (Ldpe2G) [reply projects/MAE/configs/mae_finetune.py:19]: 这里可以加一句注释，解释下这个模型是从哪里下载的

Comment (leaves-zwx) [reply projects/MAE/configs/mae_finetune.py:19]: 嗯，好的

File: libai/data/build.py [hunk 17]
```diff
@@ -348,7 +348,7 @@ def build_image_train_loader(
 def build_image_test_loader(
     dataset,
     test_batch_size,
-    sampler=LazyCall(SingleRoundSampler)(shuffle=False, drop_last=False),
+    sampler=LazyCall(SingleRoundSampler)(shuffle=True, drop_last=False),
```
[the remark this answers is not in the document]
Comment (CPFLAME) [reply libai/data/build.py:351]: 想问一下, 为什么shuffle设置为True就可以避开`persistent_workers=True`时丢失数据的问题.

[the remark this answers is not in the document]
Comment (leaves-zwx) [reply libai/data/build.py:351]: 因为不会总是丢那几个固定的数据，会换着丢。

File: projects/MAE/configs/mae_finetune.py [hunk 18]
```diff
@@ -1,20 +1,36 @@
...
 from flowvision.loss.cross_entropy import SoftTargetCrossEntropy
 
 from libai.config import LazyCall, get_config
 from .models.vit_base_patch16 import model
-from ..utils.scheduler import warmup_layerscale_cosine_lr_scheduler
+from ..utils.scheduler import (
+    warmup_layerscale_cosine_lr_scheduler,
+    warmup_cosine_lr_scheduler,
+)
 from ..utils.lr_decay import param_groups_lrd
```
Comment (CPFLAME) [inline projects/MAE/configs/mae_finetune.py:27]: 这里导入包要不把相对路径都换成绝对路径吧, 

之前黄威那边遇到过这种相对路径的config, 没有办法被pickle打包的问题

Comment (leaves-zwx) [reply projects/MAE/configs/mae_finetune.py:27]: 好的

**Review by Ldpe2G — APPROVED** [review]

File: projects/MAE/configs/mae_finetune.py (hunk 18 above)
Comment (leaves-zwx) [reply projects/MAE/configs/mae_finetune.py]: 已改成 `from utils.lr_decay import param_groups_lrd`

File: projects/MAE/utils/scheduler.py [hunk 19]
```diff
@@ -17,49 +17,51 @@
...
-        return base_lr * decay_factor
+        return lr
 
     def update_lrs(self, lrs):
         self._last_lr = []
         for i, (group, lr) in enumerate(zip(self.optimizer.param_groups, lrs)):
-            if "lr_scale" in group.options:
-                group.options["lr"] = lr * group.options["lr_scale"]
+            if "lr_scale" in group:
+                group["lr"] = lr * group["lr_scale"]
```
Comment (CPFLAME) [inline projects/MAE/utils/scheduler.py:61]: 为什么这里之前是 `group.options`
现在写法换成了`group`

Comment (leaves-zwx) [reply projects/MAE/utils/scheduler.py:61]: 是等价的，这里面确实实际操作的是 options 数据结构（在 oneflow 里面），但后面为了写法上为了与 pytorch 对齐，已经做了在 param group 上调用 `__get_attr__` 和 `__set_attr__` 实际就是调用 options 的功能。

**Review by CPFLAME — APPROVED** [review]

File: libai/data/build.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (yuanms2) [reply libai/data/build.py:351]: ”换着丢“， 能不能彻底的解决这个问题呢？

[the remark this answers is not in the document]
Comment (leaves-zwx) [reply libai/data/build.py:351]: 是这可以解决的。不过目前来看，这个丢数据的问题对实际训练的影响可能不那么大，我这里优先保障模型是可以训练出来的，这也是 kelly 目前最关心的问题。最终肯定要解决这个丢数据的问题，否则后面出了 bug，就会有说不清的问题了。
